datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -290,17 +290,211 @@ impl<'a> TypeCoercionRewriter<'a> {
+                ));
+                // cast to duration
+                let expr =
+                    Expr::Cast(Cast::new(Box::new(expr), Duration(TimeUnit::Second)));
+
+                // finally cast to interval
+                Expr::Cast(Cast::new(
+                    Box::new(expr),
+                    Interval(IntervalUnit::MonthDayNano),
+                ))
+            }
+            // These might seem to be a bit convoluted, however for arrow to do date + time arithmetic
+            // date must be cast to Timestamp(Nanosecond) and time cast to Duration(Nanosecond)
+            // (they must be the same timeunit).
+            //
+            // For Time32/64 we first need to cast to an Int64, convert that to nanoseconds based
+            // on the time unit, then cast that to duration.
+            //
+            // Time + date -> timestamp or
+            (
+                Operator::Plus | Operator::Minus,
+                Time32(_) | Time64(_),
+                Duration(TimeUnit::Nanosecond),
+                Timestamp(TimeUnit::Nanosecond, None),
+            ) => {
+                let expr = match left_current_type {
+                    Time32(TimeUnit::Second) => {
+                        let ex = Expr::Cast(Cast::new(Box::new(expr), Int64));
+                        Expr::BinaryExpr(BinaryExpr::new(
+                            Box::new(ex),
+                            Operator::Multiply,
+                            Box::new(Expr::Literal(
+                                ScalarValue::Int64(Some(1_000_000_000)),
+                                None,
+                            )),
+                        ))
+                    }
+                    Time32(TimeUnit::Millisecond) => {
+                        let ex = Expr::Cast(Cast::new(Box::new(expr), Int64));
+                        Expr::BinaryExpr(BinaryExpr::new(
+                            Box::new(ex),
+                            Operator::Multiply,
+                            Box::new(Expr::Literal(
+                                ScalarValue::Int64(Some(1_000_000)),
+                                None,
+                            )),
+                        ))
+                    }
+                    Time64(TimeUnit::Microsecond) => {
+                        let ex = Expr::Cast(Cast::new(Box::new(expr), Int64));
+                        Expr::BinaryExpr(BinaryExpr::new(
+                            Box::new(ex),
+                            Operator::Multiply,
+                            Box::new(Expr::Literal(
+                                ScalarValue::Int64(Some(1_000)),
+                                None,
+                            )),
+                        ))
+                    }
+                    Time64(TimeUnit::Nanosecond) => {
+                        Expr::Cast(Cast::new(Box::new(expr), Int64))
+                    }
+                    _ => unreachable!(),
```

Comment:
```suggestion
                    _ => internal_err!("..."),
```
in case a new time unit is ever added to the enum